apps/passport-client/pages/index.tsx

```diff
@@ -462,9 +434,83 @@ async function loadInitialState(): Promise<AppState> {
 
 registerServiceWorker();
 
-const root = createRoot(document.querySelector("#root"));
-root.render(
-  <RollbarProvider>
-    <App />
-  </RollbarProvider>
-);
+interface AppStateProviderProps {
+  children: React.ReactNode;
+  initialState: AppState;
+}
+
+const stateEmitter: StateEmitter = new Emitter();
+
+const AppStateProvider: React.FC<AppStateProviderProps> = ({
+  children,
+  initialState
+}) => {
+  const [state, setState] = useState<AppState>(initialState);
+  const [lastDiff, setLastDiff] = useState<Partial<AppState>>({});
+
+  const update = useCallback(
+    (diff: Partial<AppState>): void => {
+      setState(Object.assign(state, diff));
+
+      // In a React class component, the `setState` method has a second
+      // parameter, which is a callback function that React will invoke when
+      // the state change has taken effect. `useState` does not offer the same
+      // functionality, and the recommended approach is to use a `useEffect`
+      // hook with the relevant piece of state as a dependency. The effect hook
+      // will then be invoked whenever the state changes.
+      //
+      // However, we specifically want to observe changes to the `state`
+      // variable, and the use of `Object.assign` above ensures that, from
+      // React's perspective, the object doesn't change, at least not in a way
+      // that would trigger a re-render or an effect hook to run. This is
+      // because we do not change the object's identity, only its content.
+      //
+      // So, we need to set up some other piece of state that changes whenever
+      // the state object does. Here, we track the receipt of diffs in the
+      // update method, and in the below `useEffect` hook we trigger the hook
+      // to fire whenever a new diff is received. This allows the hook to fire
+      // on state changes even though it can't track a change to the state
+      // object directly. It will then emit an event, which is what the rest of
+      // the app uses to work around the fact that it also can't track changes
+      // to the state object.
+      setLastDiff(diff);
+    },
+    [state]
+  );
+
+  useEffect(() => {
+    stateEmitter.emit(state);
+  }, [state, lastDiff]);
+
+  const actionDispatch = useCallback(
+    (action: Action): Promise<void> => {
+      return dispatch(action, state, update);
+    },
+    [state, update]
+  );
+
+  const context = useMemo(
+    () => ({
+      getState: () => state,
+      update,
+      dispatch: actionDispatch,
+      stateEmitter
+    }),
+    [actionDispatch, state, update]
+  );
+
+  return (
+    <StateContext.Provider value={context}>{children}</StateContext.Provider>
+  );
+};
+
+loadInitialState().then((initialState: AppState) => {
+  const root = createRoot(document.querySelector("#root") as Element);
+  root.render(
+    <RollbarProvider>
+      <AppStateProvider initialState={initialState}>
+        <App />
+      </AppStateProvider>
+    </RollbarProvider>
+  );
+});
```

Comment:
would like to see a `catch` here that displays the app with an error for the user